Update protobuf to v3.2.0 for MOBILE_INFERENCE compiling.

The Mobile client of Baidu needs to use protobuf v3.2.  But if using the official protobuf v3.2.0 in https://github.com/google/protobuf in PaddlePaddle, the cross-compiling for android in linux failed. The error is as follows. The https://github.com/google/protobuf/pull/3878 solved this error by adding flags to allow one to omit building libprotoc and protoc binaries, but it merged after v3.5.0.  So I maintain protobuf in my own Github and fix the v3.2 by patching the https://github.com/google/protobuf/pull/3878 .

```
/bin/sh: ./js_embed: cannot execute binary file
make[5]: *** [/home/xx/android-cross-build/build-v8/third_party/protobuf/src/extern_protobuf/src/google/protobuf/compiler/js/well_known_types_embed.cc] Error 126
make[5]: *** Deleting file `/home/xx/android-cross-build/build-v8/third_party/protobuf/src/extern_protobuf/src/google/protobuf/compiler/js/well_known_types_embed.cc'
/home/xx/android-cross-build/build2/third_party/protobuf/src/extern_protobuf/src/google/protobuf/compiler/js/js_generator.cc:(.text._ZNK6google8protobuf8compiler2js9Generator12GenerateFileERKNS2_16GeneratorOptionsEPNS0_2io7PrinterEPKNS0_14FileDescriptorE+0x46c): undefined reference to `well_known_types_js'
/home/xx/android-cross-build/build2/third_party/protobuf/src/extern_protobuf/src/google/protobuf/compiler/js/js_generator.cc:(.text._ZNK6google8protobuf8compiler2js9Generator12GenerateFileERKNS2_16GeneratorOptionsEPNS0_2io7PrinterEPKNS0_14FileDescriptorE+0x470): undefined reference to `well_known_types_js'
libprotobuf.a(common.cc.o): In function `google::protobuf::internal::DefaultLogHandler(google::protobuf::LogLevel, char const*, int, std::string const&)':
/home/xx/android-cross-build/build2/third_party/protobuf/src/extern_protobuf/src/google/protobuf/stubs/common.cc:(.text._ZN6google8protobuf8internal17DefaultLogHandlerENS0_8LogLevelEPKciRKSs[_ZN6google8protobuf8internal17DefaultLogHandlerENS0_8LogLevelEPKciRKSs]+0x188): undefined reference to `__android_log_write'
/home/xx/android-cross-build/build2/third_party/protobuf/src/extern_protobuf/src/google/protobuf/stubs/common.cc:(.text._ZN6google8protobuf8internal17DefaultLogHandlerENS0_8LogLevelEPKciRKSs[_ZN6google8protobuf8internal17DefaultLogHandlerENS0_8LogLevelEPKciRKSs]+0x1f4): undefined reference to `__android_log_write'
```

--- a/cmake/external/protobuf.cmake
+++ b/cmake/external/protobuf.cmake
@@ -188,14 +188,24 @@ FUNCTION(build_protobuf TARGET_NAME BUILD_FOR_HOST)
         SET(OPTIONAL_CACHE_ARGS "-DZLIB_ROOT:STRING=${ZLIB_ROOT}")
     ENDIF()
 
+    SET(PROTOBUF_REPO "https://github.com/google/protobuf.git")
+    SET(PROTOBUF_TAG "9f75c5aa851cd877fb0d93ccc31b8567a6706546")
+    IF(MOBILE_INFERENCE)
+        SET(PROTOBUF_REPO "https://github.com/qingqing01/protobuf.git")
+        SET(PROTOBUF_TAG "v3.2.0")
+        IF(NOT BUILD_FOR_HOST)
+            SET(OPTIONAL_ARGS ${OPTIONAL_ARGS} "-Dprotobuf_BUILD_PROTOC_BINARIES=OFF")
+        ENDIF()
+    ENDIF()
+
     ExternalProject_Add(
         ${TARGET_NAME}
         ${EXTERNAL_PROJECT_LOG_ARGS}
         PREFIX          ${PROTOBUF_SOURCES_DIR}
         UPDATE_COMMAND  ""
         DEPENDS         zlib
-        GIT_REPOSITORY  "https://github.com/google/protobuf.git"
-        GIT_TAG         "9f75c5aa851cd877fb0d93ccc31b8567a6706546"
+        GIT_REPOSITORY  ${PROTOBUF_REPO}
+        GIT_TAG         ${PROTOBUF_TAG}
         CONFIGURE_COMMAND
         ${CMAKE_COMMAND} ${PROTOBUF_SOURCES_DIR}/src/${TARGET_NAME}/cmake
             ${OPTIONAL_ARGS}
@@ -213,7 +223,11 @@ FUNCTION(build_protobuf TARGET_NAME BUILD_FOR_HOST)
     )
 ENDFUNCTION()
 
-SET(PROTOBUF_VERSION 3.1)
+IF(NOT MOBILE_INFERENCE)
+    SET(PROTOBUF_VERSION 3.1)
+ELSE()
+    SET(PROTOBUF_VERSION 3.2)
+ENDIF()
 IF(CMAKE_CROSSCOMPILING)
     build_protobuf(protobuf_host TRUE)
     LIST(APPEND external_project_dependencies protobuf_host)

--- a/cmake/external/protobuf.cmake
+++ b/cmake/external/protobuf.cmake
@@ -191,6 +191,8 @@ FUNCTION(build_protobuf TARGET_NAME BUILD_FOR_HOST)
     SET(PROTOBUF_REPO "https://github.com/google/protobuf.git")
     SET(PROTOBUF_TAG "9f75c5aa851cd877fb0d93ccc31b8567a6706546")
     IF(MOBILE_INFERENCE)
+        # The reason why the official version is not used is described in
+        # https://github.com/PaddlePaddle/Paddle/issues/6114
         SET(PROTOBUF_REPO "https://github.com/qingqing01/protobuf.git")
         SET(PROTOBUF_TAG "v3.2.0")
         IF(NOT BUILD_FOR_HOST)

